include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,383 @@
+      {
+        dominated[p].insert(q);
+      }
+      else if (Dominates<MatType>(calculatedObjectives, q, p))
+      {
+        dominationCount[p] += 1;
+      }
+    }
+
+    if (dominationCount[p] == 0)
+    {
+      ranks[p] = 0;
+      fronts[0].push_back(p);
+    }
+  }
+
+  size_t i = 0;
+
+  while (fronts[i].size() > 0)
+  {
+    std::vector<size_t> nextFront;
+
+    for (size_t p: fronts[i])
+    {
+      for (size_t q: dominated[p])
+      {
+        dominationCount[q]--;
+
+        if (dominationCount[q] == 0)
+        {
+          ranks[q] = i + 1;
+          nextFront.push_back(q);
+        }
+      }
+    }
+
+    i++;
+    fronts.push_back(nextFront);
+  }
+}
+
+//! Check if a candidate Pareto dominates another candidate
+template<typename MatType>
+inline bool NSGA2::Dominates(std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives,
+                             size_t candidateP,
+                             size_t candidateQ)
+{
+  Info << "NSGA2::NSGA2::Dominates()" << std::endl;
+
+  bool all_better_or_equal = true;
+  bool atleast_one_better = false;
+  size_t n_objectives = calculatedObjectives[0].n_elem;
+
+  for (size_t i = 0; i < n_objectives; i++)
+  {
+    if (calculatedObjectives[candidateP](i) > calculatedObjectives[candidateQ](i))
+    {
+      // p.i is worse than q.i for the i-th objective function
+      all_better_or_equal = false;
+    }
+
+    if (calculatedObjectives[candidateP](i) < calculatedObjectives[candidateQ](i))
+    {
+      // p.i is better than q.i for the i-th objective function
+      atleast_one_better = true;
+    }
+  }
+
+  return all_better_or_equal and atleast_one_better;
+}
+
+//! Assign crowding distance to the population.
+template<typename MultiobjectiveFunctionType>
+inline void NSGA2::CrowdingDistanceAssignment(const std::vector<size_t>& front,
+                                              MultiobjectiveFunctionType& objectives,
+                                              std::vector<double>& crowdingDistance)
+{
+  if (front.size() > 0)
+  {
+    for (size_t elem: front)
+    {
+      crowdingDistance[elem] = 0;
+    }
+
+    size_t fSize = front.size();
+
+    for (size_t m = 0; m < objectives.NumObjectives(); m++)
+    {
+      crowdingDistance[front[0]] = objectives.GetMaximum(m);
+      crowdingDistance[front[fSize - 1]] = objectives.GetMaximum(m);
+
+      for (size_t i = 1; i < fSize - 1 ; i++)
+      {
+        crowdingDistance[front[i]] += (crowdingDistance[front[i-1]] -
+            crowdingDistance[front[i+1]]) /
+            (objectives.GetMaximum(m) - objectives.GetMinimum(m));
+      }
+    }
+  }
+}
+
+//! Comparator for crowding distance based sorting.
+inline bool NSGA2::CrowdingOperator(size_t idxP,
+                                    size_t idxQ,
+                                    const std::vector<size_t>& ranks,
+                                    const std::vector<double>& crowdingDistance)
+{
+  if (ranks[idxP] < ranks[idxQ])
```

Comment:
Perhaps we could just `return (ranks[idxP] < ranks[idxQ])||(ranks[idxP] == ranks[idxQ] && crowdingDistance[idxP] > crowdingDistance[idxQ])` ?

Reply:
I think it is more readable if broken down. What do you say?

Reply:
I suppose that is true. Remove the extra braces though.

Reply:
Sure, will do!